### ISSUE TITLE
Support Connection to ResourceInterpretWebhook without DNS Services

### ISSUE THREAD

**Issue by lxtywypc** [karmada-io/karmada#2998]
**What would you like to be added**:
In the code of resourceinterpreter, add the logic of building a new serviceLister to initialize a new ClusterIPServiceResolver instead of DefaultServiceResolver. 

**Why is this needed**:
If we deploy karmada with Default dnsPolicy or deploy karmada in physical machine/virtual machine directly, and there is no coreDNS or other DNS services, it would not connect to resource-interpret-webhooks successfully.

**Comment by XiShanYongYe-Chang** [comment]
Hi @lxtywypc, thanks for your contribution! Have you started using Karmada?

**Comment by lxtywypc** [comment]
> Hi @lxtywypc, thanks for your contribution! Have you started using Karmada?

Yes, and I'm working for Trip.com now. We met this problem when we tried to build a customized resource-interpreter-webhook. In our opinion, karmada components should work independently without relying other DNS service.

**Comment by RainbowMango** [comment]
Can someone remind me which service the `NewDefaultServiceResolver` resolves?
https://github.com/karmada-io/karmada/blob/eda7814ffc9f69db7f1e066ad24544fc7dcc19a7/pkg/resourceinterpreter/customizedinterpreter/customized.go#L49

**Comment by XiShanYongYe-Chang** [comment]
Hi @lxtywypc, is the resolved address configured for network mounting in `MutatingWebhookConfiguration` and `ValidatingWebhookConfiguration`? 

Like this:
https://github.com/karmada-io/karmada/blob/master/artifacts/deploy/webhook-configuration.yaml#L16

**Comment by lxtywypc** [comment]
> Can someone remind me which service the `NewDefaultServiceResolver` resolves?
> 
> https://github.com/karmada-io/karmada/blob/eda7814ffc9f69db7f1e066ad24544fc7dcc19a7/pkg/resourceinterpreter/customizedinterpreter/customized.go#L49

> Hi @lxtywypc, is the resolved address configured for network mounting in `MutatingWebhookConfiguration` and `ValidatingWebhookConfiguration`?
> 
> Like this: https://github.com/karmada-io/karmada/blob/master/artifacts/deploy/webhook-configuration.yaml#L16

It resolves the service decalred in ResourceInterpreterWebhookConfiguration. Karmada-controller-manager would connect to customized resource-interpreter-webhook by the resolved result.   

Like this:  
https://github.com/karmada-io/karmada/blob/master/examples/customresourceinterpreter/webhook-configuration.yaml#L14  
But not "url", it should be "service".  
https://github.com/karmada-io/karmada/blob/c34c443d2c2a7507c12e7bb9a52fe9a2bdb97c15/vendor/k8s.io/api/admissionregistration/v1/types.go#L534

DefaultServiceResolver could only resolve it to a service URL like ```https://{{service-name}}.{{service-namespace}}.svc:443```, but it couldn't work if there was no coreDNS or other DNS service. It would work independently if the service was resolved to ClusterIP like ```https://{{service-cluster-ip}}:443```.

**Comment by RainbowMango** [comment]
> But not "url", it should be "service".

Do you mean you specify the webhook by a service? like this:

```yaml
    clientConfig:
      service:
        namespace: foo
        name: bar
```

And the `ServiceResolver` only takes effect for this case(specify webhook by a service), am I right?

**Comment by lxtywypc** [comment]
>
> > But not "url", it should be "service".
> 
> Do you mean you specify the webhook by a service? like this:
> 
> ```yaml
>     clientConfig:
>       service:
>         namespace: foo
>         name: bar
> ```
> 
> And the `ServiceResolver` only takes effect for this case(specify webhook by a service), am I right?  

I think so it does.

**Comment by RainbowMango** [comment]
OK. Thanks.

So, the [DefaultServiceResolver](https://github.com/karmada-io/karmada/blob/eda7814ffc9f69db7f1e066ad24544fc7dcc19a7/pkg/resourceinterpreter/customizedinterpreter/customized.go#L49) resolves a service to a URL with `DNS` name, like `https://ross.andromeda.svc:443`. And the `ClusterIPServiceResolver` resolves a service to a URL according to service type:
- `ClusterIP`/`LoadBalancer`/`NodePort`: net.JoinHostPort(svc.Spec.ClusterIP, fmt.Sprintf("%d", svcPort.Port))
- `ExternalName`: net.JoinHostPort(svc.Spec.ExternalName, fmt.Sprintf("%d", port))

Seems `ClusterIPServiceResolver` doesn't rely on a DNS resolver.
I have two questions for now:
1. How does the `kube-webhook` handle it, I mean which resolver it uses?
2. For this case, `no coreDNS`, if we can use the URL way instead of the service?

**Comment by lxtywypc** [comment]
> OK. Thanks.
> 
> So, the [DefaultServiceResolver](https://github.com/karmada-io/karmada/blob/eda7814ffc9f69db7f1e066ad24544fc7dcc19a7/pkg/resourceinterpreter/customizedinterpreter/customized.go#L49) resolves a service to a URL with `DNS` name, like `https://ross.andromeda.svc:443`. And the `ClusterIPServiceResolver` resolves a service to a URL according to service type:
> 
> * `ClusterIP`/`LoadBalancer`/`NodePort`: net.JoinHostPort(svc.Spec.ClusterIP, fmt.Sprintf("%d", svcPort.Port))
> * `ExternalName`: net.JoinHostPort(svc.Spec.ExternalName, fmt.Sprintf("%d", port))
> 
> Seems `ClusterIPServiceResolver` doesn't rely on a DNS resolver. I have two questions for now:
> 
> 1. How does the `kube-webhook` handle it, I mean which resolver it uses?
> 2. For this case, `no coreDNS`, if we can use the URL way instead of the service?

I'm sorry to replying late. It took a little long to read the code of kube-apiserver.
1. `kube-webhook` is accessed by `kube-apiserver`. It seems that `kube-apiserver` register a new `Webhook` first whose `clientManager` was built with `DefaultServiceResolver`   
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/plugin.go#L34  
and then initialize the webhook with the `LoopbackServiceResolver` whose core is `EndpointServiceResolver` if enable aggregator routing and is `ClusterIPServiceResolver` otherwise.  
https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-apiserver/app/server.go#L461#L473  
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubeapiserver/admission/config.go#L52  
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/server/options/admission.go#L158  

1. We deploy the `ResourceInterpretWebhook` and its service in cluster by CI. If we use the URL way, we have to deploy the service first, then login the cluster to get service IP manually, fill the configuration and then deploy it. If the service is deleted unexpectly, or its IP is changed, we have to do as above again. And if we don't want to operation cluster manually too much...  
In summary, we think it is better to use the Service way to build the configuration if the service is deployed in cluster.

**Comment by XiShanYongYe-Chang** [comment]
Hi @lxtywypc, 

For the second point, can I understand it this way: If there is no DNS server in the cluster, the service address in the URL, like https://ross.andromeda.svc:443, cannot be automatically resolved. In this case, the Service configuration mode is recommended.

**Comment by lxtywypc** [comment]
> Hi @lxtywypc,
> 
> For the second point, can I understand it this way: If there is no DNS server in the cluster, the service address in the URL, like https://ross.andromeda.svc:443, cannot be automatically resolved. In this case, the Service configuration mode is recommended.

Yes, I think it is better.
